### ISSUE TITLE
Fix optimizer pathologies in scv2-spike experiment

### ISSUE THREAD

**Issue by jgallowa07** [matsengrp/multidms#213]
# Fix optimizer pathologies in scv2-spike experiment

## Summary

The scv2-spike experiment exhibits four interconnected problems: non-convergence (6/14 models), asymmetric shift sparsity (Delta shifts zeroed out faster than BA.2), dramatic alpha/beta0 drift with increasing fusion penalty, and decreasing Delta replicate shift correlations. These are symptoms of a single root cause: the sigmoid nonlinearity creates a degenerate (alpha, beta0) manifold, and with `beta0_ridge=0` the optimizer drifts freely along it in a replicate-specific, fusionreg-dependent manner.

## Motivation

- The scv2-spike pipeline produces results that are difficult to interpret and biologically implausible (Delta showing fewer shifts than BA.2 despite being more evolutionarily distant from BA.1)
- 6/14 models fail to converge, with erratic jumps in the objective trajectory
- Replicate correlations for Delta shifts *decrease* with increasing fusion penalty (the opposite of expected behavior), undermining confidence in the shift estimates
- These issues block manuscript-quality results from the spike analysis

### The degeneracy

The model predicts functional scores as:

```
predicted_score = alpha * (sigmoid(beta0 + X @ beta) - sigmoid(beta0 + x_wt @ beta))
```

**alpha and beta0 are not independently identifiable under the sigmoid.** Shifting beta0 moves both `sigmoid(phi_variant)` and `sigmoid(phi_wt)` together. The difference changes, but alpha can compensate — a smaller sigmoid difference times a larger alpha gives the same prediction. This creates a valley of equally-good (alpha, beta0) pairs.

With `beta0_ridge = 0`, nothing constrains where on this manifold the optimizer lands.

### How this causes each problem

**Problem 1 (non-convergence, erratic jumps):** Three contributing factors:
1. The block coordinate descent cycles between calibration (alpha) and beta blocks that see different objectives. The calibration block compensates for shifts the proximal operator zeros out, creating an oscillating cycle.
2. The `scale` factor is computed once at initialization when all beta = 0 (fusion penalty = 0). As fusionreg increases, the actual objective grows relative to `scale`, making the effective lasso threshold miscalibrated and the convergence criterion unreliable.
3. The current early stopping (#206) requires all four inner blocks to hit their tolerances AND the outer objective to stabilize. Inner blocks can oscillate at their tolerance boundary due to the degeneracy, preventing early stopping even when the objective is stable.

**Problem 2 (Delta shift sparsity too high):** Delta's wildtype differs substantially from BA.1, so beta0[Delta] drifts to place its variants in a different sigmoid region. When latent phenotypes enter sigmoid saturation, `sigmoid'(phi)` becomes exponentially small, and gradients w.r.t. Delta's beta vanish. In proximal gradient descent, when the gradient is near zero the lasso threshold dominates — shifts get zeroed out because the gradient signal is too weak to resist, not because the data supports it. BA.2 stays in the sigmoid's linear region (its WT is close to BA.1), so its gradients remain strong.

**Problem 3 (alpha drops, Delta beta0 rises):** This is the degeneracy in action. As the lasso zeros out Delta shifts, beta0[Delta] repositions latent phenotypes in the sigmoid's steepest region (maximizing amplification of remaining small differences), and alpha decreases because the sigmoid now does more of the scaling work.

**Problem 4 (Delta replicate correlation decreases):** Each replicate follows a different path along the degenerate manifold (different data -> different gradients -> different drift trajectory). At high fusionreg, ~98% of Delta shifts are zero. Which mutations retain non-zero shifts depends on the sigmoid operating point, which is replicate-specific. Different replicates retain different subsets, destroying correlation.

### In scope

- **Phase 1 (one pipeline run):** Three complementary fixes targeting different aspects:
  1. Set `beta0_ridge = 10` to anchor beta0 near the reference
  2. Recompute `scale` each outer iteration so the effective lasso threshold tracks the current objective
  3. Simplify early stopping to use only `block_tol` (#206)
- Rerun the scv2-spike pipeline and evaluate all four problems

### Out of scope

- L2-on-shifts penalty variant (Phase 2 if needed)
- Shared alpha across conditions (Phase 2 if needed)
- Identity GE control experiment (Phase 2 confirmatory)
- Changes to the data preparation pipeline
- Changes to cross-validation logic

### 1. Set `beta0_ridge = 10` in spike config

**File:** `experiments/scv2-spike/config/config.yaml` (existing)

Change line 42 from `beta0_ridge: 0` to `beta0_ridge: 10`.

This anchors non-reference beta0 values toward the reference beta0, collapsing the degenerate manifold. With the scale recomputation fix (2b), `beta0_ridge = 10` constrains beta0 drift; if non-reference beta0 values still exceed 1.0 in absolute difference from the reference, increase this value. The penalty is `beta0_ridge * sum_{d != ref} (beta0[d] - beta0[ref])^2`, summing over all non-reference conditions — penalizing differences from the reference, not clamping to zero.

### 2. Recompute `scale` at each outer iteration

**File:** `multidms/jaxmodels.py` (existing)

Move the scale computation (currently line 578, computed once before the loop) inside the `for k in range(block_iters)` loop so it is recomputed at the start of each outer iteration.

```python
# Current (line 578, before the loop):
scale = abs(
    objective_total(model, data_sets, l2reg=l2reg,
                    fusionreg=fusionreg, beta0_ridge=beta0_ridge)
)

# Proposed (inside the loop, at the top of each iteration):
for k in range(block_iters):
    scale = abs(objective_total(model, data_sets, l2reg=l2reg,
                fusionreg=fusionreg, scale=1.0, beta0_ridge=beta0_ridge))
    if scale == 0:
        scale = 1.0
    ...
```

Notes:
- Pass `scale=1.0` explicitly to get the unscaled objective value. The current code omits `scale=` and relies on the default `scale=1.0` in the function signature, but inside the loop the local `scale` variable from the previous iteration would shadow this — making `scale=1.0` explicit avoids confusion.
- The `scale == 0` guard is new protective behavior not present in the current code (where scale=0 would cause division-by-zero).

### 3. Simplify early stopping (#206)

**File:** `multidms/jaxmodels.py` (existing)

Replace the 5-condition convergence check (lines 767-774) with outer-objective-only stopping. See #206 for full details. The key change:

```python
# Current
if (
    state_calibration.error < opt_calibration.tol
    and state_β0.error < opt_β0.tol
    and state_bundle.error < opt_β.tol
    and state_nonbundle.error < opt_β.tol
    and objective_error < block_tol
):
    break

# Proposed
if objective_error < block_tol:
    if verbose:
        inner_states = {
            "calibration": (state_calibration, opt_calibration),
            "β0": (state_β0, opt_β0),
            "β_nonbundle": (state_nonbundle, opt_β),
            "β_bundle": (state_bundle, opt_β),
        }
        for name, (state, opt) in inner_states.items():
            if state.error >= opt.tol:
                print(
                    f"  warning: {name} block did not converge "
                    f"(error={state.error:.2e}, tol={opt.tol:.2e})"
                )
    break
```

## Migration / Rollback Plan

No migration required. Changes are:
- A config value change (easily reverted)
- Two localized code changes in `jaxmodels.py` (easily reverted via git)
- No API changes, no new dependencies, no schema changes

## Testing Strategy

- **Existing tests:** Run `pytest --doctest-modules -vv` to verify no regressions
- **Early stopping test:** Add a test that verifies early stopping triggers when `objective_error < block_tol` even if inner solver errors exceed their tolerances
- **Pipeline evaluation (primary validation):** Rerun scv2-spike pipeline and check:
  1. **Convergence:** All 14 models converge; objective trajectories are monotonically non-increasing with no jumps exceeding 10% relative change between consecutive iterations
  2. **Sparsity symmetry:** Delta and BA.2 sparsity-vs-fusionreg curves have slopes within 2x of each other (currently Delta's slope is 2-3x steeper). Sparsity = fraction of shift parameters `(beta[d] - beta[ref])` that are exactly zero, as reported in the trajectory diagnostics.
  3. **Alpha stability:** alpha values remain in the range [3.0, 12.0] across the fusionreg grid (i.e., within 2x of the initialization value 6.0)
  4. **Replicate correlation:** Delta shift replicate correlations increase (or stay flat) with increasing fusionreg
  5. **Cross-validation:** Validation loss still has a clear minimum (using the existing 80/20 train/test split defined in the pipeline config)
- **Per-block diagnostics:** PR #212 (merged) provides per-condition alpha, beta0, theta, and sparsity trajectories — use these to confirm beta0 stays anchored and alpha is stable

## Dependencies

No new dependencies.

## Acceptance Criteria

- [ ] `beta0_ridge = 10` set in `experiments/scv2-spike/config/config.yaml`
- [ ] Scale factor recomputed each outer iteration in `jaxmodels.fit()`
- [ ] Early stopping simplified to outer `block_tol` only (#206)
- [ ] All existing tests pass
- [ ] All 14 spike models converge (monotonically non-increasing objectives, no >10% jumps)
- [ ] Delta shift sparsity within 2x of BA.2 at each fusionreg value
- [ ] Alpha values remain in [3.0, 12.0] across fusionreg grid
- [ ] Delta replicate shift correlations non-decreasing with fusionreg

## Follow-up (Phase 2, if Phase 1 insufficient)

If the above fixes don't fully resolve the problems, consider:

- **L2 on shifts:** Change L2 penalty for non-reference conditions from `l2reg * sum(beta[d]^2)` to `l2reg * sum((beta[d] - beta[ref])^2)`, eliminating the L2/fusion tension
- **Shared alpha:** Constrain all conditions to use a single alpha value
- **Identity GE control:** Rerun with `ge_type: "Identity"` to confirm the sigmoid is causal (no code changes needed)
